### DISCUSSION TITLE
Generating schema for classes without annotation

### DISCUSSION THREAD

**Discussion by anarsultanov** [smallrye/smallrye-open-api#1353] [Q&A] [answered]
Hello,

I'd like to know if it's possible to create an OpenAPI schema from JAX-RS classes that do not have the `@Path` annotation using the `smallrye-open-api-maven-plugin`.

My use case is that I implement Keycloak extensions with custom REST endpoints, for which I would like to generate documentation using the above plugin, but ran into a problem. The problem is that the plugin doesn't recognize my resource without the `@Path` annotation, but if I add the annotation, the extension won't work properly because Quarkus will try to initialize it as a bean, which results in failure due to unsatisfied dependencies.

I would greatly appreciate any advice or suggestions on how to overcome this issue.

**Comment by phillip-kruger** [comment]
@MikeEdgar ? Any thoughts ? @anarsultanov is it possible to see something ? like an app or reproducer ?

**Reply by MikeEdgar** [reply]
A small sample would be good.

> The problem is that the plugin doesn't recognize my resource without the @Path annotation, but if I add the annotation, the extension won't work properly because Quarkus will try to initialize it as a bean, which results in failure due to unsatisfied dependencies.

I'm trying to get a better idea of this situation. Do you manage the lifecycle of the object yourself? That is, it's not a CDI bean? I was thinking maybe it could be possible to leverage https://github.com/quarkusio/quarkus/issues/16218, but that probably is a bad idea :laughing:

**Reply by anarsultanov** [reply]
@phillip-kruger @MikeEdgar thanks for the quick reply, here is this extension: https://github.com/anarsultanov/keycloak-multi-tenancy
Currently, when running `mvn package` command, correct OpenAPI schema is generated in the `docs/` directory.  This is due to the fact that I have added the `@Path` annotation to the `dev.sultanov.keycloak.multitenancy.resource.TenantsResource` class.  However, when deploying the extension to Keycloak, which is powered by Quarkus, it attempts to initialize this class, resulting in a failure with the error:
```
ERROR: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.keycloak.models.RealmModel and qualifiers [@Default]
	- java member: dev.sultanov.keycloak.multitenancy.resource.TenantsResource():realm
	- declared on CLASS bean [types=[dev.sultanov.keycloak.multitenancy.resource.TenantsResource, java.lang.Object], qualifiers=[@Default, @Any], target=dev.sultanov.keycloak.multitenancy.resource.TenantsResource]
```
Removing the annotation resolves this issue, but then the documentation is not created (more precisely, it is created, but the necessary data is missing in it).

> Do you manage the lifecycle of the object yourself? That is, it's not a CDI bean?

The lifecycle of this object is managed by Keycloak SPI (in this case, `dev.sultanov.keycloak.multitenancy.resource.TenantsResourceProvider`, which implements `RealmResourceProvider`).

> maybe it could be possible to leverage https://github.com/quarkusio/quarkus/issues/16218, but that probably is a bad idea 

I think it might work, but only until the bug is fixed :smile:

**Reply by MikeEdgar** [reply]
@anarsultanov , can you keep the `@Path` on `TenantsResource` and add `@ApplicationScoped`, remove the constructor argument, look-up the managed bean in `RealmResourceProvider`, and set the `realm` before calling `setup`? Maybe it will work if you let Quarkus manage that class's lifecycle, and you execute your setup process on the managed bean...

**Reply by anarsultanov** [reply]
@MikeEdgar won't modifying the state of an application scoped bean in a multi-threaded environment potentially cause concurrency issues?  Additionally, this could expose these endpoints on an incorrect path, since in the case of the Keycloak SPI, the path is determined by the realm and provider identifier.

Maybe there is a way to introduce some custom annotation and implement an extension for the plugin to make it properly identify the desired class without affecting the functionality (something similar to `@Schema` used for representation classes)?

**Reply by MikeEdgar** [reply]
> won't modifying the state of an application scoped bean in a multi-threaded environment potentially cause concurrency issues? 

I think it would be OK in this case as long as the state you are initializing is only setup from your SPI implementation class and the JAX-RS setup does not require that state to be set already.

> Additionally, this could expose these endpoints on an incorrect path, since in the case of the Keycloak SPI, the path is determined by the realm and provider identifier.

This is good to know. It would definitely mean that anything the annotation scanner finds would be placed in the OpenAPI document under an incorrect path and would require use of an `OASFilter` to alter it to the correct path for the realm/provider.

> Maybe there is a way to introduce some custom annotation and implement an extension for the plugin to make it properly identify the desired class without affecting the functionality (something similar to `@Schema` used for representation classes)?

This may be an option, maybe something like `@SmallRyeOpenAPIPath(value = "/your/path")`.

**Reply by anarsultanov** [reply]
> would be placed in the OpenAPI document under an incorrect path 

This is not a problem for me, I solve it by setting an empty path in annotation and specifying correct path in the "servers". But I afraid that quarkus will also expose the endpoints on the path specified in annotation.

> This may be an option, maybe something like @SmallRyeOpenAPIPath(value = "/your/path")

Something like this will definitely solve my issue. 
I'd be happy to contribute to its implementation, however, I would appreciate guidance on where to begin, as I am not familiar with the codebase.

**Reply by MikeEdgar** [reply]
@anarsultanov , I think the simplest would be to add that annotation in the `core` project in `io.smallrye.openapi.api.annotations` (a new package). You would then add its name to the JAX-RS array used to find path elements [1] and add an associated test (or tests) to the JAX-RS extension module.

[1] https://github.com/smallrye/smallrye-open-api/blob/a64b6af6156337f0a2dbc4ab9d7d559fee991817/extension-jaxrs/src/main/java/io/smallrye/openapi/jaxrs/JaxRsConstants.java#L25-L27

**Reply by anarsultanov** [reply]
Thanks @MikeEdgar, I've created a small PR introducing this annotation: https://github.com/smallrye/smallrye-open-api/pull/1365
I wasn't sure how many test cases I needed to cover, given that it can only be applied at the class level, so I added just one simple test.

**Reply by codespearhead** [reply, marked as answer]
That PR [1] was closed in favor of [2], which was merged on `2023-03-01T00:33:19Z`.

@anarsultanov Can you [mark this comment as the answer](https://docs.github.com/en/discussions/managing-discussions-for-your-community/moderating-discussions#marking-a-comment-as-an-answer) so it doesn't show up as unanswered in the GitHub discussions tab?

[1] https://github.com/smallrye/smallrye-open-api/pull/1365
[2] https://github.com/smallrye/smallrye-open-api/pull/1369

**Reply by tsaarni** [reply]
I was looking to document a Keycloak REST API extension and ended up here. I couldn’t find this feature documented anywhere, but I discovered how to use it thanks to [this snippet](https://github.com/anarsultanov/keycloak-multi-tenancy/blob/dee7730355f09cfe8bd6aff84f3b49f58769490d/pom.xml#L180-L188) in @anarsultanov 's project

```xml
            <plugin>
                <groupId>io.smallrye</groupId>
                <artifactId>smallrye-open-api-maven-plugin</artifactId>
                <executions>
                    <execution>
                        <phase>compile</phase>
                        <goals>
                            <goal>generate-schema</goal>
                        </goals>
                    </execution>
                </executions>
                <configuration>
                    <outputDirectory>${project.basedir}/docs</outputDirectory>
                    <infoTitle>${project.name}</infoTitle>
                    <infoVersion>${project.version}</infoVersion>
                    <servers>{{keycloakUrl}}/auth/realms/{{realmName}}</servers>
                    <scanResourceClasses>
                        <dev.sultanov.keycloak.multitenancy.resource.TenantsResource>/tenants</dev.sultanov.keycloak.multitenancy.resource.TenantsResource>
                    </scanResourceClasses>
                </configuration>
            </plugin>
```

The resource class `dev.sultanov.keycloak.multitenancy.resource.TenantsResource` doesn't use the `@Path` annotation, but its base path is overwritten as `/tenants`.

Thank you for this feature @anarsultanov!
